### PR TITLE
feat: ColorPicker

### PR DESCRIPTION
**Pull request by RedJue**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   New Component ColorPicker        |
| 🇨🇳 Chinese |   新组件颜色选择器        |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### Summary
🎨🧪📝

This pull request adds a new ColorPicker component to the ant-design library, which allows users to select and edit colors in different formats. It includes the component implementation, tests, demos, and documentation files. The component uses the RcColorPicker component as a base, and extends it with some custom logic, styles, and subcomponents. The component also provides a hook to manage the color state, and exports some types for external use.

> _`ColorPicker` shines_
> _A rainbow of props and tests_
> _Spring of components_

### Walkthrough
*  Add the `ColorPicker` component and its documentation ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-be92dd92152a780003d48ddc95ec6716367d327f9f890f1221b84614aaa8168fR1-R155), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-819859f3d30fc8ec12a8ef1b55ac1d75b4b1d600eb1cee989efd5af57255b560R1-R50), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-9d1ff106a032e484c8f982279c4e1295d23f69904a26e020e99310c446e93249R1-R49), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-3e9304b2ad2c5648fca4d4743d6d5af269c770f48ddbb29c57c0d901bb516e45R1-R6))
*  Implement the `ExPanel` component to extend the `RcColorPicker` panel with a `ColorDataBar` component ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-916c5ee69775c07c3ecb3505f6741803dfbbfc0a93b037af1353e447ec7e41baR1-R21), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-9c149a12b5866edb427b9066bd7b0ee0b0091c56142077e14fea6b2f989eaad8R1-R180))
*  Define the `Color` interface and the `ColorFactory` class to wrap the `RcColor` class and provide custom methods and properties ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-b05107f431c3fce3a8573df8da8022dbb0a697537ed9ad4b8e674affd5a7c98cR1-R45), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-9576ead765b69ad9edf8f148171300a9ac514ef80167a23b0ecd7e0704439647R1-R39))
*  Define the `ColorFormatEnum` type and the `ColorFormat`, `PressetsItem`, and `ColorPickerBaseProps` types for the props and internal components of the `ColorPicker` component ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-df67a3a012b615f532d14092a82649a7bf5d7d65773c192377235693e7fd26e0R1-R6), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-83dba3026e7dc41ad64d3a2f4d484eeb085f0c3798dc2316fbb90d23520fca6eR1-R16))
*  Implement the `ColorHexInput`, `ColorPlaceholder`, and `ColorSteppers` components to render and edit the color values in different formats ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-3ea70913f2f5e38335a70b730457daaa3827d75907311df6df24234f55c31304R1-R45), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-836e9e6d03a02a5cdc05b24461a55f692e4dbada93c8b07b3ab4ca5cc5337a07R1-R31), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-3fd68486b864ca56f3159c41082c985d14d24297b54639ea2213edd38c7074c0R1-R59))
*  Add the basic usage demo of the `ColorPicker` component and its markdown file ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-c9f8c74dc86457938053a55bd0dc68780cb138694afe083e6a5510e26a76f32fR1-R13), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-f126c1c22aaeef1ad8e26e6282e140fd953ec34821c205171d80f78d2de8c51cR1-R7))
*  Add the test files for the `ColorPicker` component and its demos, using the testing library and image snapshots ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-aa06460351e8630f305cf6509e8137984e0f167ab2560da22874033189234fb4R1-R10), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-c9cb9f0026b304e411b8b52dd8bfe190b1e37b96b368edca6856c89c54f9fdeaR1-R5), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-de0044e4fa03be11c72b7e52b4a26a86e5563ea6331cf7c47b17b32657db4bbeR1-R3), [link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-028bcdae3bd0c10a9e65ed01c106e4f0045ad92e288354ba40a77734eb651b1dR1-R3))
*  Add the style index file and the style hooks and tokens for the `ColorPicker` component and its subcomponents ([link](https://github.com/ant-design/ant-design/pull/41481/files?diff=unified&w=0#diff-1b0771e050d88aebe327dae413b7f31f0797bcee1515f3c6bb2cd890ba0929d0R1-R275))

### CODE REVIEW

File: components/color-picker/interface.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import type { ReactNode } from 'react';
+import type { Color } from './color';
+import type { ColorPickerProps } from './ColorPicker';
+import type { ColorFormatEnum } from './enum';
+
+export type ColorFormat = keyof typeof ColorFormatEnum;
```
Comment (MadCcc) [inline components/color-picker/interface.ts:6]: 这个用字面量就好了

File: components/color-picker/ColorPicker.tsx [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    borderRadius: token.borderRadius,
+  });
+
+  const getPlacement = () => {
+    if (!placement) {
+      return direction === 'rtl' ? 'bottomRight' : 'bottomLeft';
+    }
+
+    return placement;
+  };
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:89]: 改成 useMemo 吧

File: components/color-picker/ColorPicker.tsx [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+  const changeHandle = (data: Color) => {
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:91]:
```suggestion
  const handleChange = (data: Color) => {
```

File: components/color-picker/ColorPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    const color: Color = generateColor(data);
+    if (value) {
+      onChange?.(color, color.toHexString());
+    } else {
+      setColorValue(color);
+    }
+  };
+
+  const openChangeHandle = (openData: boolean) => {
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:100]:
```suggestion
  const handleOpenChange = (openData: boolean) => {
```

Comment (MadCcc) [reply components/color-picker/ColorPicker.tsx:100]: 这里看起来可以直接透穿吧

Comment (RedJue) [reply components/color-picker/ColorPicker.tsx:100]: 嗯 这边漏了 可以直接透下去 之前的open的逻辑也有点冗余 可以优化一下

Comment (RedJue) [reply components/color-picker/ColorPicker.tsx:100]: 好像不行，需要获取到 open 的状态来控制占位颜色的激活状态，需要有一个内部状态在非受控场景下。

**Comment by afc163** [comment]
看上去有很多无关的 commit

**Comment by RedJue** [comment]
> 看上去有很多无关的 commit

同步了一下 feature 分支

**Comment by afc163** [comment]
图片
不太对。

**Comment by RedJue** [comment]
> 图片
> 
> 不太对。

是不是因为 **rebase** 的时候解决冲突提交导致的？

**Comment by afc163** [comment]
rebase 之后没有 push --force 导致的。

**Comment by RedJue** [comment]
> rebase 之后没有 push --force 导致的。

ok 我

> rebase 之后没有 push --force 导致的。

需要重新处理一下么

**Comment by yoyo837** [comment]
重新rebae 然后 -f

**Comment by RedJue** [comment]
> 重新rebae 然后 -f

ok

File: components/color-picker/ColorPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  };
+
+  return wrapSSR(
+    <RcColorPicker
+      value={colorValue.toHsb()}
+      classNames={{
+        popup: hashId,
+      }}
+      style={{
+        popup: style,
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:150]: style 应该给 trigger，这里应该是 styles.popup.

File: components/color-picker/ColorPicker.tsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    offset: token.marginXXS,
+    arrowWidth: arrow ? token.sizePopupArrow : 0,
+    borderRadius: token.borderRadius,
+  });
+
+  useEffect(() => {
+    setPopupOpen(open);
+  }, [open]);
+
+  const currentlacement = useMemo(() => {
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:88]:
```suggestion
  const currentplacement = useMemo(() => {
```

File: components/color-picker/ColorPicker.tsx [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    defaultValue,
+  });
+  const [popupOpen, setPopupOpen] = useState(open);
+  const [clearColor, setClearColor] = useState(false);
+
+  const prefixCls = getPrefixCls('color-picker', customizePrefixCls);
+  const rootPrefixCls = getPrefixCls();
+
+  const [wrapSSR, hashId] = useStyle(prefixCls);
+  const mergeCls = classNames(`${prefixCls}-root`, className, rootClassName, hashId);
```
Comment (MadCcc) [inline components/color-picker/ColorPicker.tsx:75]: root 这个 class 应该不需要。另外 rootClassName 是也需要给 popup 的

File: components/color-picker/components/ColorClear.tsx [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import type { FC } from 'react';
+import React from 'react';
+import type { ColorPickerBaseProps } from '../interface';
+import { generateColor } from '../util';
+
+interface ColorClearProps extends ColorPickerBaseProps {}
```
Comment (MadCcc) [inline components/color-picker/components/ColorClear.tsx:6]: 这里直接 pick

File: components/color-picker/components/ColorClear.tsx [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+const ColorClear: FC<ColorClearProps> = ({ prefixCls, color, updateColor, updateClearColor }) => {
+  const handleClick = () => {
+    const hsba = color.toHsb();
+    hsba.a = 0;
+    updateColor?.(generateColor(hsba));
+    updateClearColor?.(true);
```
Comment (MadCcc) [inline components/color-picker/components/ColorClear.tsx:13]: 这个不对，antd 里应该有一个 Color 对象，不应该转成 hsb 再转回去

File: components/color-picker/components/ColorDataBar.tsx [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+import { generateColor, getAlphaColor } from '../util';
+
+import ColorHexInput from './ColorHexInput';
+import ColorSteppers from './ColorSteppers';
+
+interface ColorDataBarProps extends ColorPickerBaseProps {
+  onChange?: () => void;
+}
+
+const ColorDataBar: FC<ColorDataBarProps> = ({
```
Comment (MadCcc) [inline components/color-picker/components/ColorDataBar.tsx:17]: 每一种类型的 input 都封装一下，像 ColorHexInput 一样，这里应该做成可以直接替换的部件。
命名叫 ColorInput 比较好

File: components/color-picker/components/ColorSteppers.tsx [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  // Update step value
+  useEffect(() => {
+    if (isNumber(value)) {
+      setStepValue(value);
+    }
+  }, [value]);
+
+  return (
+    <div className={ColorSteppersPrefixCls}>
+      <div className={`${ColorSteppersPrefixCls}-input`}>
```
Comment (MadCcc) [inline components/color-picker/components/ColorSteppers.tsx:39]: 两层有点多了，看看是不是可以直接给 InputNumber 加 class

File: components/color-picker/demo/format.tsx [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        <Space>
+          <Col>
+            <ColorPicker format="hex" value={colorHex} onChange={(value) => setColorHex(value)} />
+          </Col>
+          <Col>
+            HEX: <span>{colorHex.toHexString()}</span>
+          </Col>
+        </Space>
+      </Row>
+      <Row align={'middle'}>
```
Comment (MadCcc) [inline components/color-picker/demo/format.tsx:23]:
```suggestion
      <Row align="middle">
```
其他的也改改

File: components/color-picker/enum.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+/* eslint-disable import/prefer-default-export */
+export enum ColorFormatEnum {
```
Comment (MadCcc) [inline components/color-picker/enum.ts:2]: 用字面量，一般不用 Enum

File: package.json [hunk 14]
```diff
@@ -113,6 +113,7 @@
     "@ant-design/react-slick": "~1.0.0",
     "@babel/runtime": "^7.18.3",
     "@ctrl/tinycolor": "^3.6.0",
+    "@rc-component/color-picker": "^1.0.0-alpha.1",
```
Comment (yoyo837) [inline package.json:116]: 稳定后才改`~`吗?

Comment (RedJue) [reply package.json:116]: 嗯 等那边完善完再发正式版

File: components/color-picker/components/ColorClear.tsx (hunk 9 above)
Comment (RedJue) [reply components/color-picker/components/ColorClear.tsx:13]: 新版已经改过了，这部分还是老代码😥

File: docs/react/practical-projects.en-US.md [hunk 15]
```diff
@@ -165,7 +165,7 @@ Assuming we have agreed on an API interface with the backend developers, we can
 Create a `mock` directory and add a new `products.ts` file to this directory with the following code.
 
 ```ts
-import { defineMock } from "umi";
+import { defineMock } from 'umi';
```
Comment (afc163) [inline docs/react/practical-projects.en-US.md]: 为啥会改这个？

Comment (afc163) [reply docs/react/practical-projects.en-US.md]: 还是有很多无关 commits

**Comment by afc163** [comment]
- [ ] 整个弹层为啥不直接复用 Popover，这样动画方向和箭头都不用重复实现了。

File: docs/react/practical-projects.en-US.md (hunk 15 above)
Comment (RedJue) [reply docs/react/practical-projects.en-US.md]: 我的问题，`rebase` 的时候没有 同步 fork 的 `feature` 分支

**Comment by RedJue** [comment]
> * Popover

`rc-color-picker` 用的 `trigger` 实现的弹层，`antd` 这里封装也是延用

**Pull request closed without merging**